images/storage-network-discoverer/src/pkg/discoverer/discoverer.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discoverer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/netip"
+	"os"
+	"slices"
+	"strings"
+	"time"
+
+	"k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"storage-network-controller/internal/config"
+	"storage-network-controller/internal/logger"
+	"storage-network-controller/internal/utils"
+	"storage-network-controller/pkg/cache"
+)
+
+type discoveredIPs []string
+
+const (
+	// this type will be added to node's status.addresses field
+	RVStorageIPType = "SDSRVStorageIP"
+)
+
+var DiscoveryCache *cache.TTLCache[string, string]
+
+func DiscoveryLoop(ctx context.Context, cfg *config.Options, mgr manager.Manager) error {
+	// just syntactic sugar for logger
+	log := logger.FromContext(ctx)
+
+	storageNetworks, err := parseCIDRs(cfg.StorageNetworkCIDR)
+	if err != nil {
+		log.Error(err, "Cannot parse storage-network-cidr")
+		return err
+	}
+
+	myNodeName := os.Getenv("NODE_NAME")
+	if myNodeName == "" {
+		return errors.New("cannot get node name because no NODE_NAME env variable")
+	}
+
+	// we cannot do `mgr.GetClient()` because its require a cache, but we do not want cache feature
+	// so, create our own kubeclient WITHOUT cache
+	cl, err := client.New(mgr.GetConfig(), client.Options{})
+	if err != nil {
+		log.Error(err, "Cannot create k8s client")
+		return err
+	}
+
+	log.Info("Discoverer settings:")
+	log.Info(fmt.Sprintf("* Log level: %s", cfg.Loglevel))
+	log.Info(fmt.Sprintf("* Discovery every %d seconds", cfg.DiscoverySec))
+	log.Info(fmt.Sprintf("* Cache founded IP(s) for %d seconds", cfg.CacheTTLSec))
+	log.Info(fmt.Sprintf("* Storage network CIDRs: %s", storageNetworks))
+	log.Info(fmt.Sprintf("* I discovery IPs on node %s", myNodeName))
+
+	// create a new DiscoveryCache with TTL (item expiring) capabilities
+	DiscoveryCache = cache.NewTTL[string, string](ctx)
+
+	// discoverer loop
+	for {
+		select {
+		case <-ctx.Done():
+			// do nothing in case of context canceling and just return from the loop
+			return nil
```

Comment:
Maybe will be cleaner to have it this way? So it will be clear that function failed and why. Not sure if it will need other changes.

```suggestion
			return ctx.Err()
```

Reply:
Just thoughts. Feel free to resolve.

Reply:
From [here](https://stackoverflow.com/questions/67439974/ctx-err-vs-ctx-done-chan-closed-check-is-there-a-difference): 

> That doesn't sound right. ctx.Err returns an error for cancelled contexts as well.

So it seems to me that there will be a `context.Canceled` error in `ctx.Err()` which returned as result of `DiscoveryLoop` function (with an exit code > 0 and Kubernetes will restart that "failed" pod instead of just terminate it successfully). It seems impractical to do a separate check for `context.Canceled` error in `runController()`. So I decided to leave `return nil` here.